File: Algorithm/binary_search.py
def binary_search(arr, left, right, key):
    mid = int((left+right)/2)
    if left == right:
        if arr[mid] == key:
            return ("Found",mid)
        else:
            return ("Not Found",left-1, right)
    else:
        if arr[mid] == key:
            return ("Found", mid)
        elif arr[mid] > key:
            right = mid-1
            return binary_search(arr, left, right, key)
        else:
            left = mid+1
            return binary_search(arr, left, right, key)


def tmp(arr, left, right, key):
    # key不在起始值范围之内
    if arr[left] > key:
        return ("Not Found","~", left)
    elif arr[right] < key:
        return ("Not Found",right, "~")
    # key值在起始值范围之内
    else:
        return binary_search(arr, left, right, key)


if __name__ == '__main__':
    # arr = [1,2,3,5]
    arr = [1, 2]
    print(tmp(arr, 0, len(arr)-1, 1.5))

File: Algorithm/binary_search.cpp
#include <iostream>
using namespace std;

int *result; // 结果
int mid;     // 中间值的index

int *binary_search(int *arr, int left, int right, int key) {
    mid = (left + right) / 2;
    if (left > right) {
        result = new int[2]{right, left};
        return result;
    }
    if (arr[mid] > key)
        return binary_search(arr, left, mid - 1, key);
    else if (arr[mid] < key)
        return binary_search(arr, mid + 1, right, key);
    else {
        result = new int[2]{mid, mid};
        return result;
    }
}

int main() {
    int *arr = new int[4]{1, 2, 4, 5};
    int *tmp = binary_search(arr, 0, 3, 3);
    cout << "(" << tmp[0] << "," << tmp[1] << ")" << endl;
}